File: go/evaluate_reverse_polish_notation.go
package main

import "strconv"

func evalRPN(tokens []string) int {
	stack := []int{}
	for _, tok := range tokens {
		n := len(stack)
		switch tok {
		case "+":
			stack[n-2] += stack[n-1]
			stack = stack[:n-1]
		case "-":
			stack[n-2] -= stack[n-1]
			stack = stack[:n-1]
		case "*":
			stack[n-2] *= stack[n-1]
			stack = stack[:n-1]
		case "/":
			stack[n-2] /= stack[n-1]
			stack = stack[:n-1]
		default:
			v, _ := strconv.Atoi(tok)
			stack = append(stack, v)
		}
	}
	return stack[0]
}

File: c/evaluate_reverse_polish_notation.c
int evalRPN(char **tokens, int tokensSize) {
	int stack[10000];
	int k = 0;
	for (int i = 0; i < tokensSize; i++) {
		char *tok = tokens[i];
		switch (tok[0]) {
		case '*':
			stack[k - 2] *= stack[k - 1];
			k--;
			break;
		case '/':
			stack[k - 2] /= stack[k - 1];
			k--;
			break;
		case '+':
			stack[k - 2] += stack[k - 1];
			k--;
			break;
		case '-':
			if (tok[1] == '\0') {
				stack[k - 2] -= stack[k - 1];
				k--;
				break;
			}
		default:
			stack[k] = atoi(tok);
			k++;
		}
	}
	return stack[0];
}
